qaframework-bdd-tests/src/test/java/org/openmrs/contrib/qaframework/automation/AddUserAndAddRolesToUserSteps.java

```diff
@@ -0,0 +1,151 @@
+    }
+
+    private void reLoginAsAdmin() throws InterruptedException {
+        goToLoginPage().loginAsAdmin();
+    }
+    
+    private void fillInRoleModules(Map<String, Integer> roleModules) {
+        roleModules.put("roleStrings.Anonymous", 0);
+        roleModules.put("roleStrings.Authenticated", 0);
+        roleModules.put("roleStrings.Organizational:SystemAdministrator", 2);
+        roleModules.put("roleStrings.SystemDeveloper", 9);
+        roleModules.put("roleStrings.Application:AdministersSystem", 1);
+        roleModules.put("roleStrings.Organizational:Doctor", 3);
+        roleModules.put("roleStrings.Organizational:Nurse", 4);
+        roleModules.put("roleStrings.Organizational:RegistrationClerk", 3);
+    }
+	
+    @Before(RunTest.HOOK.SELENIUM_ADD_USER_AND_ADD_ROLES_TO_USER)
+    public void visitHomePage() {
+        initiateWithLogin();
+    }
+
+    @After(RunTest.HOOK.SELENIUM_ADD_USER_AND_ADD_ROLES_TO_USER)
+    public void destroy() throws InterruptedException {
+        quit();
+    }
+	
+    @Given("a user clicks on system administration app")
+    public void clickOnSystemAdministrationApp() {
+        manageUserPage = homePage.goToAdministration().clickOnManageUsers();
+    }
+	
+    @Then("the system loads manage users page")
+    public void systemLoadsManageUsersPage() {
+        assertTrue(textExists("User Management"));
+    }
+	
+    @When("a user fills in person details")
+    public void fillInPersonDetails() {
+        manageUserPage.clickOnAddUser().createNewPerson().fillInPersonName("Super", "Nurse", NURSE_USERNAME, NURSE_PASSWORD);
+    }
+	
+    @When("a user clicks the add user link")
+    public void clickOnAddUserLink() {
+        addEditUserPage = manageUserPage.clickOnAddUser();
+        addEditUserPage.createNewPerson();
+    }
+	
+    @And("a user assigns roles to the created user")
+    public void assignRolesToUser() throws InterruptedException {
+        Map<String, Integer> roleModules = new HashMap<>();
+        fillInRoleModules(roleModules);
+        String oldRole = null;
+        for (Entry<String, Integer> role : roleModules.entrySet()) {
+             manageUserPage.assignRolesToUser(oldRole, role.getKey(), NURSE_USERNAME);
+
+        reLoginAsUser();
+        if (homePage.numberOfAppsPresent() != role.getValue()) {
+            throw new AssertionError("role " + role + " doesn't have matching number of accessible applications: "
+                		+ "should be: " + role.getValue() + "is: " + homePage.numberOfAppsPresent());
+        }
+
+        reLoginAsAdmin();
+        oldRole = role.getKey();
+        homePage.goToAdministration().clickOnManageUsers();
+        }
+    }
+	
+    @Then("the system loads the create new user page")
+    public void systemLoadsCreateNewUserPage() {
+        assertTrue(textExists("Add/Edit User"));
+    }
+	
+    @And("a user enters the details of the user")
+    public void enterUserDetails() {
+        addEditUserPage.saveUser();
```

Comment:
won't this test fail if the system is trying to save an existing user?

Reply:
no i believe it wont fail.

Reply:
Take a close look at the method name!